src/python/nnfusion/runtime.py

```diff
@@ -1,71 +1,140 @@
                               input_names=input_names,
                               output_names=output_names)  # , opset_version=11)
-            codegen(os.path.join(self.workdir, "stencil.onnx"),
-                    " ".join([
-                        "-f onnx",
-                        "-fextern_result_memory=1",
-                        "-fkernel_tuning_steps=" + str(self.tuning_step),
-                        "-fir_based_fusion=1",
-                        "-fkernel_fusion_level=0",
-                        # "-fantares_mode=1",
-                        # f"-fantares_codegen_server={self.codegen_server}",
-                        "-fblockfusion_level=0"]),
-                    self.workdir)
-            modify_nnfusion_rt(rt_dir)
-            build(rt_dir)
-        if rebuild:
-            build(rt_dir)
-        self.executor = Executor(rt_dir)
-        self.input_name = self.executor.get_inputs()[0].name
-        self.output_name = self.executor.get_outputs()[0].name
-
-    def run(self, input, output):
-        in_dict = {}
-        out_dict = {}
-        if isinstance(input, list):
-            # assert len(self.executor.get_inputs()) == len(input)
-            for i in range(len(input)):
-                in_dict[self.executor.get_inputs(
-                )[i].name] = cast_pytorch_tensor(input[i])
-        else:
-            in_dict = {self.input_name: cast_pytorch_tensor(input)}
-        if isinstance(output, list):
-            # assert len(self.executor.get_outputs()) == len(output)
-            for i in range(len(output)):
-                out_dict[self.executor.get_outputs(
-                )[i].name] = cast_pytorch_tensor(output[i])
-        else:
-            out_dict = {self.output_name: cast_pytorch_tensor(output)}
 
+        def check_if_need_build():
+            if not os.path.exists(self.onnx_path):
+                return True
+
+            # Compare onnx file to check if modified
+            with tempfile.TemporaryDirectory(dir=self.workdir) as tmp:
+                temp_onnx_path = os.path.join(tmp, "temp.onnx")
+                export_onnx(temp_onnx_path)
+
+                if not filecmp.cmp(temp_onnx_path, self.onnx_path):
+                    # Replace the original to avoid exporting onnx twice
+                    os.remove(self.onnx_path)
+                    os.link(temp_onnx_path, self.onnx_path)
+                    return True
+            return False
+
+        def do_compile():
+            if not os.path.exists(self.onnx_path):
+                export_onnx(self.onnx_path)
+
+            codegen(self.onnx_path, self.compile_flag, self.workdir)
+            modify_nnfusion_rt(self.rt_dir)
+            build(self.rt_dir)
+
+        if force_build and os.path.exists(self.onnx_path):
+            os.remove(self.onnx_path)
+
+        if check_if_need_build():
+            do_compile()
+
+        self._init_executor(self.rt_dir, device=inputs[0].device)
+
+    def _init_executor(self, rt_dir, device):
```

Comment:
Could you move it into Executor? 

Executor is designed to wrap ".so", so please avoid direct operation on ".so" out of Executor.

Reply:
Hi, thanks for the feedback.
Since Executor is expected to initialize corresponding device in the constructor, should I pass a boolean variable (decide whether to reserve) and the device into Executor constructor and make the executor hold the tensor?
Thanks.

Reply:
I think you can distinguish them by checking whether 'get_workspace_size' exists (or whether return value = 0, depends on the implementation without function_codegen flag). 

Avoid adding additional parameters when creating Executor because it should be self-explanatory.

Reply:
Ok.